### PR TITLE
Array operations pt1

### PR DESCRIPTION
**Pull request by KTGypsy**
## What:

* Implementation of map, reduce & filter operations.

## Why:

* To extend operations sets and complete the fundaments of our engine.

## Example:

* https://jsonlogic.com/operations.html#map-reduce-and-filter

### CODE REVIEW

File: src/commonTest/kotlin/operations/array/ReduceTest.kt [hunk 1]
```diff
@@ -0,0 +1,117 @@
+package operations.array
+
+import JsonLogicEngine
+import TestInput
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.datatest.withData
+import io.kotest.matchers.shouldBe
+
+class ReduceTest : FunSpec({
```
[the remark this answers is not in the document]
Comment (PaulinaSadowskaAllegro) [reply src/commonTest/kotlin/operations/array/ReduceTest.kt:9]: also when in the registry is some value that cannot be parsed to double

**Review by PaulinaSadowskaAllegro — CHANGES_REQUESTED** [review]

File: src/commonMain/kotlin/CommonJsonLogicEngine.kt [hunk 2]
```diff
@@ -61,7 +63,14 @@ internal class CommonJsonLogicEngine : JsonLogicEngine {
         DoubleNegation.operation,
         And.operation,
         Or.operation,
-        If.operation,
+        If.operation
+    )
+
+    private val selfEvaluatingOperations: Map<String, (Any?, Any?) -> Any?> = mapOf(
+        // array
+        operations.array.Map.operation,
```
[the remark this answers is not in the document]
Comment (KTGypsy) [reply src/commonMain/kotlin/CommonJsonLogicEngine.kt:71]: It conflicts with map as collection usage so i write it like this. Do you have any any idea other than type aliasing?

File: src/commonMain/kotlin/operations/array/unwrap/EvaluatingUnwrapStrategy.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+internal interface EvaluatingUnwrapStrategy {
+    fun evaluate(expression: Map<String, Any?>, data: Any?) = JsonLogicEngine.instance.evaluate(expression, data)
+
+    fun unwrapOperationData(expression: List<Any?>, data: Any?) =
+        (expression.firstOrNull().unwrapOperationData(data) as? List<*>)
+
+    private fun Any?.unwrapOperationData(data: Any?): Any? = when {
+        this is List<*> -> map { it.unwrapOperationData(data) }
+        isExpression(this) -> evaluate(this as Map<String, Any?>, data)
```
[the remark this answers is not in the document]
Comment (KTGypsy) [reply src/commonMain/kotlin/operations/array/unwrap/EvaluatingUnwrapStrategy.kt:14]: isExpression() function is meant to check it

File: src/commonMain/kotlin/operations/array/unwrap/EvaluatingUnwrapStrategy.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package operations.array.unwrap
+
+import JsonLogicEngine
+import kotlin.collections.Map
+
+internal interface EvaluatingUnwrapStrategy {
+    fun evaluate(expression: Map<String, Any?>, data: Any?) = JsonLogicEngine.instance.evaluate(expression, data)
```
[the remark this answers is not in the document]
Comment (KTGypsy) [reply src/commonMain/kotlin/operations/array/unwrap/EvaluatingUnwrapStrategy.kt]: I extracted evaluation logic 👍🏽

File: src/commonMain/kotlin/JsonLogicEngine.kt [hunk 5]
```diff
@@ -6,5 +34,46 @@ interface JsonLogicEngine {
...
+            NotStrictEquals.operation,
+            Negation.operation,
+            DoubleNegation.operation,
+            And.operation,
+            Or.operation,
+            If.operation
+        )
+        internal val selfEvaluatingOperations: Map<String, (Any?, Any?) -> Any?> = mapOf(
+            // array
+            operations.array.Map.operation,
```
Comment (PaulinaSadowskaAllegro) [inline src/commonMain/kotlin/JsonLogicEngine.kt:74]: missing import

**Review by PaulinaSadowskaAllegro — DISMISSED** [review]

**Review by mkrogulskiallegro2 — APPROVED** [review]

**Pull request merged**